content/application-service-api.md

```diff
@@ -375,9 +376,19 @@ service's `user` namespaces. If the parameter is missing, the homeserver
 is to assume the application service intends to act as the user implied
 by the `sender_localpart` property of the registration.
 
+{{% added-in v="1.17" %}} Application services MAY similarly masquerade
+as a specific device ID belonging the user ID through use of the `device_id`
+query string parameter on the request. If the given device ID is not known
+to belong to the user, the server will return a 400 `M_UNKNOWN_DEVICE` error.
+If no `user_id` is supplied, the `device_id` MUST belong to the user implied
+by the `sender_localpart` property of the application service's registration.
+If no `device_id` is supplied, the homeserver is to assume the request is
+being made without a device ID and will fail to complete operations which
+require a device ID (such as uploading one-time keys).
```

Comment:
Should we also specify what happens when this is used on requests where the device id is not used?

Reply:
That should be specified implicitly: the request continues as normal.

Reply:
From how I read the current text the server would have to check the device id even when it is unneeded.
It should probably include a sentence to explicitly allow the server to check the device id in this case.

Something like:

For requests that do not use the device id the server MAY omit to check the `device_id` argument.

Reply:
If a device ID is passed, it should be verified regardless of whether the endpoint needs it

Reply:
> What is the point of that, wouldn't that just be a (mostly) useless database query on the server?

The purpose of the MSC was to mirror the `user_id` validation rather than try to define new semantics. A future MSC can remove the requirement to validate the device ID, though it may be valuable to keep around to maintain the "last seen IP" information within a homeserver.

Reply:
I guess its fine if this mirrors the user id behavior.
I still thing it might be good to allow not checking for lightweight servers, but its probably not that much of a difference anyways.